File: Turquoise/Manager/TQUserInfoManager.m

#import "TQUserInfoManager.h"

#import "KeychainWrapper.h"

#define kUserNameKey ((NSString *)kSecAttrAccount)
#define kPasswordKey ((NSString *)kSecValueData)
#define kFullNameKey ((NSString *)kSecAttrLabel)
#define kEmailKey ((NSString *)kSecAttrService)

static NSString *const kGroupsKey = @"userInfo.groups";

@interface TQUserInfoManager ()

// this actually should be an NSSet, but NSUserDefaults can't store NSSets.
//@property(nonatomic, copy, readonly) NSMutableDictionary<NSString *, NSNumber *> *subscribedGroups;

@end

@implementation TQUserInfoManager {
  KeychainWrapper *_keychain;
  NSMutableDictionary<NSString *, NSNumber *> *_subscribedGroups;
}

+ (instancetype)sharedInstance {
  static TQUserInfoManager *theManager = nil;
  static dispatch_once_t onceToken;
  dispatch_once(&onceToken, ^{
    theManager = [[self alloc] init];
    theManager->_keychain = [[KeychainWrapper alloc] init];

    theManager->_subscribedGroups =
        [[[NSUserDefaults standardUserDefaults] objectForKey:kGroupsKey] mutableCopy];
    if (!theManager->_subscribedGroups) {
      theManager->_subscribedGroups = [@{ @"metu.ceng.test" : @1 } mutableCopy];
      [[NSUserDefaults standardUserDefaults] setObject:theManager->_subscribedGroups
                                                forKey:kGroupsKey];
    }
  });

  return theManager;
}

#pragma mark - Properties

- (NSString *)userName {
  return [self userInfoValueForKey:kUserNameKey];
}

- (void)setUserName:(NSString *)userName {
  [self setUserInfoValue:userName forKey:kUserNameKey];
}

- (NSString *)password {
  return [self userInfoValueForKey:kPasswordKey];
}

- (void)setPassword:(NSString *)password {
  [self setUserInfoValue:password forKey:kPasswordKey];
}

- (NSString *)fullName {
  return [self userInfoValueForKey:kFullNameKey];
}

- (void)setFullName:(NSString *)fullName {
  [self setUserInfoValue:fullName forKey:kFullNameKey];
}

- (NSString *)email {
  return [self userInfoValueForKey:kEmailKey];
}

- (void)setEmail:(NSString *)email {
  [self setUserInfoValue:email forKey:kEmailKey];
}

- (NSArray<NSString *> *)sortedSubscribedGroupIds {
  return [_subscribedGroups.allKeys sortedArrayUsingSelector:@selector(localizedCaseInsensitiveCompare:)];
}

#pragma mark -

- (void)resetUserCredentials {
  [_keychain resetKeychainItem];
}

- (BOOL)isSubscribedToGroup:(TQNNTPGroup *)group {
  if (!group.groupId) {
    return NO;
  }
  return (_subscribedGroups[group.groupId] != nil);
}

- (void)subscribeToGroup:(TQNNTPGroup *)group {
  if (!group.groupId) {
    return;
  }

  _subscribedGroups[group.groupId] = @1;
  [[NSUserDefaults standardUserDefaults] setObject:_subscribedGroups forKey:kGroupsKey];
  [[NSNotificationCenter defaultCenter] postNotificationName:kUserSubscriptionsChangedNotification
                                                      object:self];
  NSLog(@"[INFO] Subscribed to group '%@'", group.groupId);
}

- (void)unsubscribeFromGroup:(TQNNTPGroup *)group {
  if (!group.groupId) {
    return;
  }

  [_subscribedGroups removeObjectForKey:group.groupId];
  [[NSUserDefaults standardUserDefaults] setObject:_subscribedGroups forKey:kGroupsKey];
  [[NSNotificationCenter defaultCenter] postNotificationName:kUserSubscriptionsChangedNotification
                                                      object:self];
  NSLog(@"[INFO] Unsubscribed from group '%@'", group.groupId);
}

- (id)userInfoValueForKey:(NSString *)key {
  NSAssert(key, @"Cannot get value, key is nil" );
  return [_keychain myObjectForKey:key];
}

- (void)setUserInfoValue:(id)value forKey:(NSString *)key {
  NSAssert(key, @"Cannot set value, key is nil" );
  NSAssert(value, @"Cannot set nil value in Keychain" );
  [_keychain mySetObject:value forKey:key];
}

@end

File: Turquoise/Manager/TQUserInfoManager.swift
import Foundation
import Security

class TQUserInfoManager {
  let kUserNameKey = kSecAttrAccount as String
  let kPasswordKey = kSecValueData as String
  let kFullNameKey = kSecAttrLabel as String
  let kEmailKey = kSecAttrService as String

  let userSubscriptionsDidChangeNotification = Notification.Name("userSubscriptionsDidChangeNotification")
  let userDidLogoutNotification = Notification.Name("userDidLogoutNotification")
  let groupsKey = "userInfo.groups"

  static let sharedInstance = TQUserInfoManager()

  var userName: String? {
    get {
      return self.userInfoValueFor(key: kUserNameKey) as! String?
    }
    set {
      self.setUserInfo(value: newValue as NSCoding?, forKey: kUserNameKey)
    }
  }

  var password: String? {
    get {
      return self.userInfoValueFor(key: kPasswordKey) as! String?
    }
    set {
      self.setUserInfo(value: newValue as NSCoding?, forKey: kPasswordKey)
    }
  }

  var fullName: String? {
    get {
      return self.userInfoValueFor(key: kFullNameKey) as! String?
    }
    set {
      self.setUserInfo(value: newValue as NSCoding?, forKey: kFullNameKey)
    }
  }

  var email: String? {
    get {
      return self.userInfoValueFor(key: kEmailKey) as! String?
    }
    set {
      self.setUserInfo(value: newValue as NSCoding?, forKey: kEmailKey)
    }
  }
  var sortedSubscribedGroupIds: [String]? {
    get {
      guard let subscribedGroups = self.subscribedGroups else {
        return nil
      }

      return subscribedGroups.keys.sorted { (string1, string2) -> Bool in
        string1.localizedCompare(string2) == .orderedAscending
      }
    }
  }

  private var keychain: KeychainWrapper!
  var subscribedGroups: [String : Any]?

  init() {
    let serviceName = Bundle.main.bundleIdentifier ?? "AyranKeychainService"
    self.keychain = KeychainWrapper(serviceName: serviceName)

    self.subscribedGroups = UserDefaults.standard.object(forKey: self.groupsKey) as! [String : Any]?
    if self.subscribedGroups == nil {
      self.subscribedGroups = [ "metu.ceng.test" : 1 ]
      UserDefaults.standard.set(self.subscribedGroups, forKey:self.groupsKey)
    }
  }

  // MARK: -

  func resetUserCredentials() {
    // TODO: it's probably a good idea to check the return value here.
    _ = self.keychain.removeAllKeys()
    self.subscribedGroups = [ "metu.ceng.test" : 1 ]
    UserDefaults.standard.set(self.subscribedGroups, forKey: self.groupsKey)
    NotificationCenter.default.post(name: self.userDidLogoutNotification, object: self)
  }

  func isSubscribedTo(group: TQNNTPGroup) -> Bool {
    if self.subscribedGroups == nil {
      // TODO: error
      return false
    }
    return self.subscribedGroups![group.groupId] != nil
  }

  func subscribeTo(group: TQNNTPGroup) {
    if self.subscribedGroups == nil {
      // TODO: error
      return
    }

    self.subscribedGroups![group.groupId] = 1
    UserDefaults.standard.set(self.subscribedGroups, forKey: self.groupsKey)
    NotificationCenter.default.post(name: self.userSubscriptionsDidChangeNotification, object: self)
    printInfo("Subscribed to group '\(group.groupId)'")
  }

  func unsubscribeFrom(group: TQNNTPGroup) {
    if self.subscribedGroups == nil {
      // TODO: error
      return
    }

    self.subscribedGroups!.removeValue(forKey: group.groupId)
    UserDefaults.standard.set(self.subscribedGroups!, forKey: self.groupsKey)
    NotificationCenter.default.post(name: self.userSubscriptionsDidChangeNotification, object: self)
    printInfo("Unsubscribed from group '\(group.groupId)'")
  }

  func userInfoValueFor(key: String) -> Any? {
    return self.keychain.object(forKey: key)
  }

  func setUserInfo(value: NSCoding?, forKey key: String) {
    if let value = value {
      self.keychain.set(value, forKey: key)
    } else {
      self.keychain.removeObject(forKey: key)
    }
  }



}
